src/common/src/types/jsonb.rs

```diff
@@ -380,6 +383,96 @@ impl<'a> JsonbRef<'a> {
+                | DataType::Varchar
+                | DataType::List(_)
+                | DataType::Struct(_)
+        ) {
+            return Err(format!("cannot cast jsonb to {ty}"));
+        }
+        if self.0.as_null().is_some() {
+            return Ok(None);
+        }
+        Ok(Some(match ty {
+            DataType::Jsonb => ScalarImpl::Jsonb(self.into()),
+            DataType::Boolean => ScalarImpl::Bool(self.as_bool()?),
+            DataType::Int16 => ScalarImpl::Int16(self.as_number()?.try_into()?),
+            DataType::Int32 => ScalarImpl::Int32(self.as_number()?.try_into()?),
+            DataType::Int64 => ScalarImpl::Int64(self.as_number()?.try_into()?),
+            DataType::Float32 => ScalarImpl::Float32(self.as_number()?.try_into()?),
+            DataType::Float64 => ScalarImpl::Float64(self.as_number()?),
+            DataType::Varchar => ScalarImpl::Utf8(self.force_string().into()),
+            DataType::List(t) => ScalarImpl::List(self.to_list(t)?),
+            DataType::Struct(s) => ScalarImpl::Struct(self.to_struct(s)?),
+            _ => unreachable!(),
+        }))
+    }
+
+    /// Convert the jsonb value to a list value.
+    pub fn to_list(self, elem_type: &DataType) -> Result<ListValue, String> {
+        let array = self
+            .0
+            .as_array()
+            .ok_or_else(|| format!("expected JSON array, but found {self}"))?;
+        let mut builder = elem_type.create_array_builder(array.len());
+        for v in array.iter() {
+            builder.append(Self(v).to_datum(elem_type)?);
+        }
+        Ok(ListValue::new(builder.finish()))
+    }
+
+    /// Convert the jsonb value to a struct value.
+    pub fn to_struct(self, ty: &StructType) -> Result<StructValue, String> {
+        let object = self.0.as_object().ok_or_else(|| {
+            format!(
+                "cannot call populate_composite on a jsonb {}",
+                self.type_name()
+            )
+        })?;
+        let mut fields = Vec::with_capacity(ty.len());
+        for (name, ty) in ty.iter() {
+            let datum = match object.get(name) {
+                Some(v) => Self(v).to_datum(ty)?,
+                None => None,
+            };
+            fields.push(datum);
+        }
+        Ok(StructValue::new(fields))
+    }
+
+    /// Expands the top-level JSON object to a row having the struct type of the `base` argument.
+    pub fn populate_struct(
+        self,
+        ty: &StructType,
+        base: StructRef<'_>,
+    ) -> Result<StructValue, String> {
+        let object = self.0.as_object().ok_or_else(|| {
+            format!(
+                "cannot call populate_composite on a jsonb {}",
+                self.type_name()
+            )
+        })?;
+        let mut fields = Vec::with_capacity(ty.len());
+        for ((name, ty), base_field) in ty.iter().zip_eq_debug(base.iter_fields_ref()) {
+            let datum = match object.get(name) {
+                Some(v) => Self(v).to_datum(ty)?,
```

Comment:
Is the fallback-to-base behavior supposed to happen recursively? I just tried the following in PostgreSQL:
```
test=# create table pt (x int, y int);
CREATE TABLE
test=# create table sg (a pt, b pt);
CREATE TABLE
test=# select jsonb_populate_record(row(row(1, 2)::pt, row(3, 4)::pt)::sg, '{"a": {"x":9},"b":{"y":17}}');
 jsonb_populate_record 
-----------------------
 ("(9,2)","(3,17)")
```

But in this PR:
```
dev=> select jsonb_populate_record(row(row(1, 2), row(3, 4))::struct<a struct<x int, y int>, b struct<x int, y int>>, '{"a": {"x":9},"b":{"y":17}}');
 jsonb_populate_record 
-----------------------
 ("(9,)","(,17)")
(1 row)
```

Reply:
It seems so. I should fix this.